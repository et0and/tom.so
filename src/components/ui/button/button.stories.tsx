import * as React from "react";
import { Button } from "./button";
import { StoryFn, Meta } from "@storybook/react";

// Define the metadata for the component
export default {
  title: "Components/Button",
  component: Button,
  argTypes: {
    variant: {
      control: "select",
      options: ["default", "destructive", "secondary", "link"],
      description: "The visual style of the button",
    },
    size: {
      control: "select",
      options: ["default", "sm", "lg", "icon"],
      description: "The size of the button",
    },
    children: {
      control: "text",
      description: "The text content of the button",
    },
    onClick: {
      action: "clicked",
      description: "Click event handler",
    },
    disabled: {
      control: "boolean",
      description: "Whether the button is disabled",
    },
  },
} as Meta<typeof Button>;

// Create a template for the component
const Template: StoryFn<typeof Button> = (args) => <Button {...args} />;

// Create different stories for each variant and size
export const Default = Template.bind({});
Default.args = {
  children: "Default Button",
};

export const Destructive = Template.bind({});
Destructive.args = {
  children: "Destructive Button",
  variant: "destructive",
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: "Secondary Button",
  variant: "secondary",
};

export const Link = Template.bind({});
Link.args = {
  children: "Link Button",
  variant: "link",
};

export const Small = Template.bind({});
Small.args = {
  children: "Small Button",
  size: "sm",
};

export const Large = Template.bind({});
Large.args = {
  children: "Large Button",
  size: "lg",
};

export const Icon = Template.bind({});
Icon.args = {
  children: (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <path d="M23 19a2 2 0 0 1-2 2H3a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h4l2-3h6l2 3h4a2 2 0 0 1 2 2v11z"></path>
      <circle cx="12" cy="13" r="4"></circle>
    </svg>
  ),
  size: "icon",
  "aria-label": "Icon Button", // Important for accessibility
};

export const Disabled = Template.bind({});
Disabled.args = {
  children: "Disabled Button",
  disabled: true,
};

export const AsChild = Template.bind({});
AsChild.args = {
  children: <a href="#">Link as Child</a>,
  asChild: true,
};
