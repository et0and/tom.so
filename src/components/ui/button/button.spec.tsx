import React from "react";
import { render } from "@testing-library/react";
import { Button, ButtonProps } from "./button";

describe("<Button/>", () => {
  const defaultProps: ButtonProps = {
    children: "Click me",
  };

  it("renders default button correctly", () => {
    const { asFragment } = render(<Button {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button with different variants", () => {
    const variants: ButtonProps["variant"][] = [
      "default",
      "destructive",
      "outline",
      "secondary",
      "ghost",
      "link",
    ];
    variants.forEach((variant) => {
      const { asFragment } = render(
        <Button {...defaultProps} variant={variant} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("renders button with different sizes", () => {
    const sizes: ButtonProps["size"][] = ["default", "sm", "lg", "icon"];
    sizes.forEach((size) => {
      const { asFragment } = render(<Button {...defaultProps} size={size} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("renders button as child", () => {
    const { asFragment } = render(
      <Button {...defaultProps} asChild>
        <a href="https://example.com">Link Button</a>
      </Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button with custom className", () => {
    const { asFragment } = render(
      <Button {...defaultProps} className="custom-class" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders disabled button", () => {
    const { asFragment } = render(<Button {...defaultProps} disabled />);
    expect(asFragment()).toMatchSnapshot();
  });
});
