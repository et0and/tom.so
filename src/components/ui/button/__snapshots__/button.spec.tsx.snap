// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> renders button as child 1`] = `
<DocumentFragment>
  <a
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
    href="https://example.com"
  >
    Link Button
  </a>
</DocumentFragment>
`;

exports[`<Button/> renders button with custom className 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2 custom-class"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different sizes 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different sizes 2`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-9 rounded-md px-3"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different sizes 3`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-11 rounded-md px-8"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different sizes 4`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 w-10"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 2`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-red-500 text-gray-50 hover:bg-red-500/90 dark:bg-red-900 dark:text-gray-50 dark:hover:bg-red-900/90 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 3`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 border border-gray-200 bg-white hover:bg-gray-100 hover:text-gray-900 dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 4`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-100 text-gray-900 hover:bg-gray-100/80 dark:bg-gray-800 dark:text-gray-50 dark:hover:bg-gray-800/80 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 5`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 hover:bg-gray-100 hover:text-gray-900 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders button with different variants 6`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 text-gray-900 underline-offset-4 hover:underline h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders default button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`<Button/> renders disabled button 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-gray-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-gray-950 dark:focus-visible:ring-gray-300 bg-gray-900 text-gray-50 hover:bg-gray-900/90 dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 h-10 px-4 py-2"
    disabled=""
  >
    Click me
  </button>
</DocumentFragment>
`;
